### PR TITLE
Add support for pandas1.0.0 nullable string/bool/int

### PR DESCRIPTION
**Pull request by cpitchford**
Pandas 1.0.0 introduces a collection of nullable dtypes.
This initial cut updates 0.3.3 with support for reading and writing
these types:

    * BooleanDtype
    * StringDtype
    * Int8/16/32/64Dtype
    * UInt8/16/32/64Dtype

This cut requires pandas >= 1.0.0 and makes no attempt at supporting older There are no tests yet
It relies on string comparisions on dtypes (since numpy can't compare its dtypes
against pandas dtypes)

It's a first cut for feedback and I expect to add testing and bring in line
with dask/fastparquet style

There remains a defect that read strings use None rather than pd.NA

```python
import pandas as pd
import fastparquet

nullint = pd.Series([10, 20, None, 40], dtype=pd.Int64Dtype())
nullstr = pd.Series(["aaa", None, "ccc", "ddd"], dtype=pd.StringDtype())
nullbool = pd.Series([True, None, False, True], dtype=pd.BooleanDtype())

sdf = pd.DataFrame({"number": nullint, "strings": nullstr, "booleans": nullbool})

print(sdf)

fastparquet.write("testing.parq", sdf)

rpq = fastparquet.ParquetFile("testing.parq")
rdf = rpq.to_pandas()

print(rdf)
print(rdf.isna())
```

Gives result:
```
   number strings  booleans
0      10     aaa      True
1      20    <NA>      <NA>
2    <NA>     ccc     False
3      40     ddd      True
   number strings  booleans
0      10     aaa      True
1      20    None      <NA>
2    <NA>     ccc     False
3      40     ddd      True
   number  strings  booleans
0   False    False     False
1   False     True      True
2    True    False     False
3   False    False     False
```

### CODE REVIEW

**Comment by martindurant** [comment]
Thank you, this looks like it will be very useful.

@TomAugspurger , wouldn't mind your thoughts here, if you have time.

Fixes #465 

Some general thoughts:
- if the column type is int(optional), then probably the dataframe type should be nullable int whether or not there are any nulls in the dataset; vice-versa for a dataframe being written, there should be no need to count nulls if the type explicitly allows nulls
- obviously, we will require the new pandas version
- should have some tests to/from nullable column types


Hah, it might take an annoying amount of tedium to correct tests:
```
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64
```
(I assume one here is straight int64, the other nullable int64)

**Comment by martindurant** [comment]
OK, got it passing.
@TomAugspurger , would still appreciate your input.

File: fastparquet/dataframe.py [hunk 1]
```diff
@@ -94,10 +96,13 @@ def cat(col):
             df[six.text_type(col)] = Categorical([], categories=cat(col),
                                                  fastpath=True)
         else:
-            if hasattr(t, 'base'):
+            if hasattr(t, 'base') and t.base is not None:
                 # funky pandas not-dtype
                 t = t.base
-            d = np.empty(0, dtype=t)
+            if hasattr(t,'na_value'):
```
Comment (TomAugspurger) [inline fastparquet/dataframe.py]: `pd.api.extensions.is_extension_array_dtype(t)` is probably the best way to check this.

File: fastparquet/dataframe.py [hunk 2]
```diff
@@ -181,6 +186,9 @@ def set_cats(values, i=i, col=col, **kwargs):
             new_block = block.make_block_same_class(
                 type(block.values)(values, dtype=block.values.dtype)
             )
+        elif hasattr(block.values.dtype,'na_value'):
+            values = pd.array([None]*size, dtype=block.values.dtype)
```
Comment (TomAugspurger) [inline fastparquet/dataframe.py]: Possibly safer to use `[dtype.na_value]`? That seems a bit safer than assuming that the dtype will coerce None to the NA value.

**Comment by martindurant** [comment]
> all reads use nullable types by default

That is the idea - is that not a key feature of pandas 1?

We could more simply cope with writing dfs which had these types, but then they would not round-trip; maybe it's something that should be in the file footer "pandas metadata", but that has not been codified for these new types.

**Comment by TomAugspurger** [comment]
They're all optional. Pandas will still infer, e.g. `pd.Series([1, 2])` as
NumPy int64. and `pd.Series([1, 2, None])` as NumPy's float64.

**Comment by martindurant** [comment]
Hm, but they do map much better to the pandas type system, so it would make a lot of sense to use them. 

As an aside, is it possible to have a float column with nulls, *and* non-null nan values?

**Comment by TomAugspurger** [comment]
No, pandas doesn't currently have a nullable FloatArray. That may happen in
the future. When it does, we likely will
allow only NA, and not NaN, in it.

File: fastparquet/dataframe.py (hunk 1 above)
Comment (cpitchford) [reply fastparquet/dataframe.py]: I've used pd.api.types.is_extension_array_dtype(t)

File: fastparquet/dataframe.py (hunk 2 above)
Comment (cpitchford) [reply fastparquet/dataframe.py]: As suggested, I'm using block.values.dtype.na_value

**Comment by martindurant** [comment]
As per @TomAugspurger 's comment, perhaps we should split this into two PRs: one that copes with those extension types at write time, if we come across them, and another one (to be merged only later) that creates new extension types on reading for all relevant "optional" columns.

**Comment by ssantana-ns** [comment]
@martindurant is there any info on when the write support will drop. Ran into this problem and we require fastparquet because for some reason dask pyarrow timestamps dont work with Athena (although pandas pyarrow does) and fastparquet has `times=int96` which fixes that issue

**Comment by martindurant** [comment]
@ssantana-ns : unfortunately, there is no plan at the moment. It would be great if someone were to rewrite this PR to split out the write part (and deal with conflicts). I'd be happy to review and release it, if it did appear.

**Comment by martindurant** [comment]
I'm going to close this - but the second part could be resubmitted as a PR, following #525

**Pull request closed without merging**